python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_reexecute.py

```diff
@@ -139,3 +147,157 @@ def test_reexec_from_parent_3():
             "2": 40,
         }
+
+
+@op
+def echo(x):
+    return x
+
+
+@op
+def adder(ls: List[int]) -> int:
+    return sum(ls)
+
+
+@op(out=DynamicOut())
+def dynamic_op():
+    for i in range(10):
+        yield DynamicOutput(value=i, mapping_key=str(i))
+
+
+def dynamic_with_optional_output_job():
+    @op(out=DynamicOut(is_required=False))
+    def dynamic_optional_output_op(context):
+        for i in range(10):
+            if (
+                context.pipeline_run.parent_run_id
+                and i % 2 == 0  # re-execution run skipped odd numbers
+                or not context.pipeline_run.parent_run_id
+                and i % 2 == 1  # root run skipped even numbers
+            ):
+                yield DynamicOutput(value=i, mapping_key=str(i))
+
+    @job(resource_defs={"io_manager": fs_io_manager})
+    def _dynamic_with_optional_output_job():
+        dynamic_results = dynamic_optional_output_op().map(echo)
+        adder(dynamic_results.collect())
+
+    return _dynamic_with_optional_output_job
+
+
+def test_reexec_dynamic_with_optional_output_job_1():
+    with instance_for_test() as instance:
+        result = dynamic_with_optional_output_job().execute_in_process(instance=instance)
+
+        # re-execute all
+        re_result = reexecute_pipeline(
+            reconstructable(dynamic_with_optional_output_job),
+            parent_run_id=result.run_id,
+            instance=instance,
+        )
+        assert re_result.success
+        assert re_result.output_for_solid("adder") == sum([i for i in range(10) if i % 2 == 0])
+
+
+def test_reexec_dynamic_with_optional_output_job_2():
+    with instance_for_test() as instance:
+        result = dynamic_with_optional_output_job().execute_in_process(instance=instance)
+
+        # re-execute the step where the source yielded an output
+        re_result = reexecute_pipeline(
+            reconstructable(dynamic_with_optional_output_job),
+            parent_run_id=result.run_id,
+            instance=instance,
+            step_selection=["echo[1]"],
+        )
+        assert re_result.success
+        assert re_result.result_for_solid("echo").output_value() == {
+            "1": 1,
+        }
+
+
+def test_reexec_dynamic_with_optional_output_job_3():
+    with instance_for_test() as instance:
+        result = dynamic_with_optional_output_job().execute_in_process(instance=instance)
+
+        # re-execute the step where the source did not yield
+        # -> error because the dynamic step wont exist in execution plan
+        with pytest.raises(
+            DagsterExecutionStepNotFoundError,
+            match=r"Step selection refers to unknown step: echo\[0\]",
+        ):
+            reexecute_pipeline(
+                reconstructable(dynamic_with_optional_output_job),
+                parent_run_id=result.run_id,
+                instance=instance,
+                step_selection=["echo[0]"],
```

Comment:
no `echo[0]` shown up in dagit:
![image](https://user-images.githubusercontent.com/4531914/142155156-44d01fdb-f20f-4acc-b0ec-e7498e5050ea.png)